### PR TITLE
Rewrite in Typescript

### PR DESCRIPTION
**Pull request by robiot**
Everything should work. I have tested it and it starts express, launches chrome logs perfectly. I have not seen it successfully find something, but it should work. 
Also added ESlint and  a more lightweight logger.

### CODE REVIEW

**Comment by robiot** [comment]
Dunno why it includes commits from my old pr lol

File: .eslintrc.json [hunk 1]
```diff
@@ -0,0 +1,19 @@
+{
+    "parser": "@typescript-eslint/parser",
+    "parserOptions": {
+        "ecmaVersion": 2021
+    },
+    "extends": ["plugin:lvksh/recommended"],
```
Comment (kalkih) [inline .eslintrc.json:6]: If we want eslint we should extend something convetional / standardized

File: .prettierrc.json [hunk 2]
```diff
@@ -1 +1,4 @@
-{}
+{
+    "tabWidth": 4,
+    "singleQuote": false
```
Comment (kalkih) [inline .prettierrc.json:3]: Skip these changes

File: src/index.ts [hunk 3]
```diff
@@ -0,0 +1,227 @@
+import yargs from "yargs";
+import { hideBin } from "yargs/helpers";
+const { argv } = yargs(hideBin(process.argv));
+
+import { Locations } from "./locations";
+import { logger } from "./logger";
+import { readConfig } from "./readConfig";
+// eslint-disable-next-line unused-imports/no-unused-imports
+import { server } from "./server";
```
Comment (kalkih) [inline src/index.ts:9]: We can simply import and skip the from here
```suggestion
import "./server";
```

File: src/index.ts [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+        .map((value) => ({ value, sort: Math.random() }))
+        .sort((a, b) => a.sort - b.sort)
+        .map(({ value }) => value);
+}
+
+function randomDate(start: Date, end: Date) {
+    const date = new Date(
+        start.getTime() + Math.random() * (end.getTime() - start.getTime())
+    );
+
```
Comment (kalkih) [inline src/index.ts:193]: Minor thing, but we don't need spacing between everything, makes it harder to read in the end.

File: src/readConfig.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
+import fs from "node:fs";
+import path from "node:path";
+
+import { Location } from "./locations";
+import { logger } from "./logger";
+
+export type ConfigT = {
```
Comment (kalkih) [inline src/readConfig.ts:7]: Skip the T here
```suggestion
export type Config = {
```

File: src/services/bookingService.ts [hunk 6]
```diff
@@ -0,0 +1,415 @@
+/* eslint-disable unicorn/error-message */
```
Comment (kalkih) [inline src/services/bookingService.ts:1]: What's this for?

File: src/services/bookingService.ts [hunk 7]
```diff
@@ -0,0 +1,415 @@
...
+                        ...current,
+                    }),
+                    { Next: "Nästa" }
+                )
+        );
+        logger.success("Residency set");
+    }
+
+    // Can fix type later
+    async postRequest(body: any) {
```
Comment (kalkih) [inline src/services/bookingService.ts:110]: Could type this as `Record<string, string | number | boolean | undefined>` or something, better than nothing

File: src/services/bookingService.ts [hunk 8]
```diff
@@ -0,0 +1,415 @@
...
+
+        return recovered;
+    }
+
+    // Config type can also be fixed
+    async bookSlot(
+        serviceTypeId: number,
+        timeslot: string,
+        location: string,
+        config: any
```
Comment (kalkih) [inline src/services/bookingService.ts:190]: You made a config interface, should be able to use it here.

File: src/services/bookingService.ts [hunk 9]
```diff
@@ -0,0 +1,415 @@
...
+            SearchTimeHour: 12,
+            RegionId: 0,
+            Next: "Nästa",
+            ReservedDateTime: timeslot,
+        });
+        const $ = cheerio.load(await response!.text());
+        const title = $(TITLE_SELECTOR).text();
+
+        if (title !== "Uppgifter till bokningen") {
+            const error = new Error() as CheerioError;
```
Comment (kalkih) [inline src/services/bookingService.ts:243]: This shouldn't be a `CheerioError`, we should type our own error type for this.

Comment (kalkih) [reply src/services/bookingService.ts:243]: Same for the errors in all related functions

File: src/services/bookingService.ts [hunk 10]
```diff
@@ -0,0 +1,415 @@
...
+        region.toLowerCase()
+    )}`;
+const generatePreviousUrl = (region: string) =>
+    `https://bokapass.nemoq.se/Booking/Booking/Previous/${replaceSpecialChars(
+        region.toLowerCase()
+    )}`;
+
+type CheerioError = Error & { page: CheerioAPI };
+
+const confirm_booking = "Bekräfta bokning";
```
Comment (kalkih) [inline src/services/bookingService.ts:31]:
```suggestion
const CONFIRM_BOOKING_TITLE = "Bekräfta bokning";
```

File: src/services/bookingService.ts [hunk 11]
```diff
@@ -0,0 +1,415 @@
...
+        }
+
+        if (title !== "Din bokning är nu klar") {
+            const error = new Error() as CheerioError;
+
+            error.page = $;
+            throw error;
+        }
+
+        const control_class = ".control-freetext";
```
Comment (kalkih) [inline src/services/bookingService.ts:395]: camelCase variable name

File: src/services/captchaService.ts [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+// require("puppeteer-extra-plugin-stealth/evasions/navigator.hardwareConcurrency");
+// require("puppeteer-extra-plugin-stealth/evasions/navigator.languages");
+// require("puppeteer-extra-plugin-stealth/evasions/navigator.permissions");
+// require("puppeteer-extra-plugin-stealth/evasions/navigator.plugins");
+// require("puppeteer-extra-plugin-stealth/evasions/navigator.vendor");
+// require("puppeteer-extra-plugin-stealth/evasions/navigator.webdriver");
+// require("puppeteer-extra-plugin-stealth/evasions/sourceurl");
+// require("puppeteer-extra-plugin-stealth/evasions/user-agent-override");
+// require("puppeteer-extra-plugin-stealth/evasions/webgl.vendor");
+// require("puppeteer-extra-plugin-stealth/evasions/window.outerdimensions");
```
Comment (kalkih) [inline src/services/captchaService.ts:21]: We need this for `pkg` to work (compile source into executables).

**Review by kalkih — CHANGES_REQUESTED** [review]
Great stuff!

I would prefer if we kept this PR to the ts conversion only, skip eslint & changed logger e.t.c.

This also breaks packaging of the executables.

**Pull request closed without merging**